src/tophu/multiscale.py

```diff
@@ -0,0 +1,482 @@
+        the pass-band and the highest gain in the stop-band), in decibels. (default: 40)
+
+    Returns
+    -------
+    out : numpy.ndarray
+        The output filtered and multilooked data.
+    """
+    arr = np.asanyarray(arr)
+
+    if arr.ndim != 2:
+        raise ValueError("input array must be 2-dimensional")
+    if (overhang < 0.0) or (overhang > 1.0):
+        raise ValueError("overhang must be between 0 and 1")
+    if len(downsample_factor) != 2:
+        raise ValueError("downsample factor should be a pair of ints")
+
+    def get_filter_coeffs(n: int) -> NDArray:
+        # If `n` is 1, then the subsequent multilooking step will be a no-op, so there's
+        # no need to apply a low-pass filter. In that case, we choose the filter
+        # coefficents to have a frequency response of unity.
+        if n == 1:
+            return 1.0
+
+        # Get the ratio of the multilooked data sampling rate to the original data
+        # sampling rate.
+        ratio = 1.0 / n
+
+        # Compute the normalized bandwidth of the pass-band (relative to the original
+        # sampling rate), given the specified `shape_factor` and `overhang` values.
+        bandwidth = ratio / (shape_factor * (1.0 - overhang) + overhang)
+
+        # Get low pass filter coefficients. Force the filter length to be odd in order
+        # to avoid applying a phase delay.
+        return bandpass_equiripple_filter(
+            bandwidth=bandwidth,
+            shape=shape_factor,
+            ripple=ripple,
+            attenuation=attenuation,
+            force_odd_length=True,
+        )
+
+    # Get low-pass filter coefficients for each axis to be multilooked. The combined 2-D
+    # filter is the Cartesian product of the 1-D filters for each individual axis.
+    coeffs1d = map(get_filter_coeffs, downsample_factor)
+    coeffs2d = np.outer(*coeffs1d)
+
+    # Filter the input data.
+    filtered = scipy.signal.fftconvolve(arr, coeffs2d, mode="same")
+
+    # Perform spatial averaging.
+    return multilook(filtered, downsample_factor)
+
+
+def upsample_unwrapped_phase(
+    igram_hires: NDArray[np.complexfloating],
+    igram_lores: NDArray[np.floating],
+    unwrapped_phase_lores: NDArray[np.floating],
+    conncomp_lores: NDArray[np.unsignedinteger],
+) -> NDArray[np.floating]:
+    r"""
+    Upsample an unwrapped phase field resulting from coarse unwrapping.
+
+    Each pair of wrapped and unwrapped phase values are assumed to differ only by an
+    integer number of cycles of :math:`2\pi`. In that case, we can form the upsampled
+    unwrapped phase by upsampling the difference between the low-resolution unwrapped
+    and wrapped phase as an integer number of cycles and then adding it to the
+    high-resolution wrapped phase.
+
+    Parameters
+    ----------
+    igram_hires : numpy.ndarray
+        The full-resolution interferogram. A two-dimensional array.
```

Comment:
Is this complex interferogram or only phase?
```suggestion
        The full-resolution wrapped interferogram. A two-dimensional array.
```

Reply:
It's the complex-valued interferogram, not the wrapped phase. But actually in this function I think we only need the phase of the interferogram, not the interferogram itself. So I guess it would make sense to change this interface to take the wrapped phase instead.